### PR TITLE
Implement standalone GC loading in NativeAOT

### PR DESCRIPTION
**Pull request by cshung**
This is currently a prototype - the silly naming reflects that. It probably doesn't work in the integrated settings yet, creating this PR to test only.

Highlights:
- [x] Known to work on Windows/Linux targetting 8.0 preview 7 to power ASP.Net BasicMinimalApi benchmark
- [x] Module loading code is confirmed to be not in the final binary using dead code elimination in the linker.
- [x] MSBuild integration

Lowlights: 
- [x] We have to build a NativeAOT-specific variant of the clrgc. (See https://github.com/dotnet/runtime/pull/91821 for the progress towards fixing this)
- [ ] All diagnostics events from CLRGC are not fired 
- [x] Naming of various classes
- [x] Reading module name from configuration
- [x] Fallback to default GC in the standalone mode if the configuration is not set.
- [x] Proper error reporting

@dotnet/gc

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/gcheaputilities.cpp [hunk 1]
```diff
@@ -47,6 +49,425 @@ extern "C" HRESULT GC_Initialize(
...
+};
+
+// Load status of the GC. If GC loading fails, the value of this
+// global indicates where the failure occurred.
+GC_LOAD_STATUS g_gc_load_status = GC_LOAD_STATUS_BEFORE_START;
+
+// The version of the GC that we have loaded.
+VersionInfo g_gc_version_info;
+
+class MyGCToEEInterface : public IGCToCLR {
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/gcheaputilities.cpp:70]: Why do we need to add all this to the shared file? Should this be in the clrgc.enabled.cpp file only?

Comment (cshung) [reply src/coreclr/nativeaot/Runtime/gcheaputilities.cpp:70]: Can we put a whole class in a different lib/a file? I was hoping that the linker is smart enough so that if there is no call site to construct this, it won't get into the binary.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 2]
```diff
@@ -53,6 +53,8 @@ The .NET Foundation licenses this file to you under the MIT license.
       <EventPipeName Condition="'$(EventSourceSupport)' == 'true'">libeventpipe-enabled</EventPipeName>
 
       <LinkStandardCPlusPlusLibrary Condition="'$(LinkStandardCPlusPlusLibrary)' == '' and '$(_IsiOSLikePlatform)' == 'true' and '$(InvariantGlobalization)' != 'true'">true</LinkStandardCPlusPlusLibrary>
+      <ClrgcSupportName>libclrgc-disabled</ClrgcSupportName>
+      <ClrgcSupportName Condition="'$(IlcEnableClrgc)' == 'true'">libclrgc-enabled</ClrgcSupportName>
```
Comment (jkotas) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:57]: `EnableClrgc` suggests that the GC enabled by default is not CLR GC, that does not sound right.

We call this [standalone GC in the doc](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#standalone-gc), so this setting should use the same name. Something like: `IlcStandaloneGCSupport`.

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 3]
```diff
@@ -671,6 +671,42 @@ REDHAWK_PALEXPORT char* PalCopyTCharAsChar(const TCHAR* toCopy)
...
+    {
+        return 0;
+    }
+    for (size_t i = 0; i < len; i++)
+    {
+        moduleNameWide[i] = (wchar_t)moduleName[i];
+    }
+    moduleNameWide[len] = '\0';
+    
+    HANDLE result = LoadLibraryExW(moduleNameWide, NULL, 0);
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:692]: This needs to pass in specific flags. Otherwise, it is going to be flagged during security scans.

File: src/coreclr/nativeaot/Runtime/gcheaputilities.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/gcheaputilities.cpp:70]: It would make it much easier to see all the optional code and ensure that it does not get pulled in by accident.

It would match how the optional support for eventpipe is structured. We should use the same approach for all optional native AOT components.

File: src/coreclr/nativeaot/Runtime/gcheaputilities.cpp [hunk 4]
```diff
@@ -47,6 +49,27 @@ extern "C" HRESULT GC_Initialize(
 
 #ifndef DACCESS_COMPILE
 
+// Load status of the GC. If GC loading fails, the value of this
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/gcheaputilities.cpp:52]: Should this be in clr.enabled.cpp ? It is only used in that file.

File: src/coreclr/nativeaot/Runtime/gcheaputilities.h [hunk 5]
```diff
@@ -102,4 +108,23 @@ class GCHeapUtilities {
     GCHeapUtilities() = delete;
 };
 
+#ifndef DACCESS_COMPILE
+
+enum GC_LOAD_STATUS {
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/gcheaputilities.h:113]: Should this be in clr.enabled.cpp ? It is only used in that file.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets [hunk 6]
```diff
@@ -53,6 +53,8 @@ The .NET Foundation licenses this file to you under the MIT license.
       <EventPipeName Condition="'$(EventSourceSupport)' == 'true'">libeventpipe-enabled</EventPipeName>
 
       <LinkStandardCPlusPlusLibrary Condition="'$(LinkStandardCPlusPlusLibrary)' == '' and '$(_IsiOSLikePlatform)' == 'true' and '$(InvariantGlobalization)' != 'true'">true</LinkStandardCPlusPlusLibrary>
+      <ClrgcSupportName>libclrgc-disabled</ClrgcSupportName>
+      <ClrgcSupportName Condition="'$(IlcStandaloneGCSupport)' == 'true'">libclrgc-enabled</ClrgcSupportName>
```
Comment (jkotas) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:57]:
```suggestion
      <StandaloneGCSupportName>libstandalonegc-disabled</StandaloneGCSupportName>
      <StandaloneGCSupportName Condition="'$(IlcStandaloneGCSupport)' == 'true'">libstandalonegc-enabled</StandaloneGCSupportName>
```
To match `IlcStandaloneGCSupport` name?

File: src/coreclr/nativeaot/Runtime/gcheaputilities.cpp [hunk 7]
```diff
@@ -7,6 +7,8 @@
 #include "gchandleutilities.h"
 
 #include "gceventstatus.h"
+#include "holder.h"
+#include "RhConfig.h"
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/gcheaputilities.cpp:11]: Are these needed?

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 8]
```diff
@@ -671,6 +671,42 @@ REDHAWK_PALEXPORT char* PalCopyTCharAsChar(const TCHAR* toCopy)
     return converted;
 }
 
+REDHAWK_PALEXPORT HANDLE PalLoadLibrary(const char* moduleName)
+{
+    if (moduleName == nullptr)
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:676]: Why do we need to treat null?

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 9]
```diff
@@ -671,6 +671,42 @@ REDHAWK_PALEXPORT char* PalCopyTCharAsChar(const TCHAR* toCopy)
...
+    wchar_t* moduleNameWide = new (nothrow)wchar_t[len + 1];
+    if (moduleNameWide == nullptr)
+    {
+        return 0;
+    }
+    for (size_t i = 0; i < len; i++)
+    {
+        moduleNameWide[i] = (wchar_t)moduleName[i];
+    }
+    moduleNameWide[len] = '\0';
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:690]: This should use `MultiByteToWideChar` - the string being passed is UTF-8 and this "poor man's converter" will produce broken results if the file path to the executable or the string in `DOTNET_GCName` contain non-ascii characters.

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 10]
```diff
@@ -671,6 +671,42 @@ REDHAWK_PALEXPORT char* PalCopyTCharAsChar(const TCHAR* toCopy)
...
+REDHAWK_PALEXPORT void* PalGetProcAddress(HANDLE module, const char* functionName)
+{
+    if (module == 0)
+    {
+        return nullptr;
+    }
+    if (functionName == nullptr)
+    {
+        return nullptr;
+    }
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:706]: Why do we need to handle these?

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp (hunk 8 above)
Comment (cshung) [reply src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:676]: So far, no call site will pass `null`, so it is just defensive programming. I can easily remove it.

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp (hunk 10 above)
Comment (cshung) [reply src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:706]: So far, no call site will pass `null`, so it is just defensive programming. I can easily remove it.

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp [hunk 11]
```diff
@@ -671,6 +671,33 @@ REDHAWK_PALEXPORT char* PalCopyTCharAsChar(const TCHAR* toCopy)
...
+REDHAWK_PALEXPORT HANDLE PalLoadLibrary(const char* moduleName)
+{
+    assert(moduleName);
+    size_t len = strlen(moduleName);
+    wchar_t* moduleNameWide = new (nothrow)wchar_t[len + 1];
+    if (moduleNameWide == nullptr)
+    {
+        return 0;
+    }
+    if (MultiByteToWideChar(CP_ACP, 0, moduleName, -1, moduleNameWide, (int)len) == 0)
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:683]: Should it be `CP_UTF8` or `CP_ACP`? `PalCopyTCharAsChar` that is used to create the string in the caller is `CP_UTF8`.

**Comment by cshung** [comment]
Look like something is broken after applying the code review feedback. Debugging now, meanwhile, do not merge.

**Comment by MichalStrehovsky** [comment]
> Look like something is broken after applying the code review feedback. Debugging now, meanwhile, do not merge.

Do we want a test so that the CI tests this? How is standalone GC tested on CoreCLR?

File: src/coreclr/gc/gcload.cpp [hunk 12]
```diff
@@ -82,7 +82,11 @@ GC_Initialize(
     assert(clrToGC == nullptr);
 #endif
 
-#ifndef FEATURE_NATIVEAOT // GCConfig and GCToOSInterface are initialized in PalInit
+#if defined(BUILD_AS_STANDALONE) || !defined(FEATURE_NATIVEAOT)
```
Comment (jkotas) [inline src/coreclr/gc/gcload.cpp:85]: Is this change still needed?

File: src/coreclr/nativeaot/Runtime/clrgc.disabled.cpp [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifdef _WIN32
+// this must exactly match the typedef used by windows.h
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/clrgc.disabled.cpp:5]: Can we include whatever defines HRESULT instead of duplicating the definition here?

**Comment by cshung** [comment]
> > Look like something is broken after applying the code review feedback. Debugging now, meanwhile, do not merge.
> 
> Do we want a test so that the CI tests this? How is standalone GC tested on CoreCLR?

StandaloneGC is supposed to be tested [here](https://dev.azure.com/dnceng-public/public/_build?definitionId=146) as a separate pipeline that turns it on for all tests. To my surprise, the pipeline is currently disabled. I will follow up with the @dotnet/gc team on this.

File: src/coreclr/gc/gcload.cpp [hunk 14]
```diff
@@ -82,7 +82,11 @@ GC_Initialize(
     assert(clrToGC == nullptr);
 #endif
 
-#ifndef FEATURE_NATIVEAOT // GCConfig and GCToOSInterface are initialized in PalInit
+#ifndef FEATURE_NATIVEAOT
+
+    // For NativeAOT, GCConfig and GCToOSInterface are initialized in PalInit
+    // In case of NativeAOT using clrgc, the copy inside clrgc needs to be intiailized again
```
Comment (jkotas) [inline src/coreclr/gc/gcload.cpp:88]:
```suggestion
```
This comment is not very relevant anymore (also, it has a typo).

File: src/coreclr/nativeaot/Runtime/clrgc.enabled.cpp [hunk 15]
```diff
@@ -0,0 +1,465 @@
...
+};
+
+// Load status of the GC. If GC loading fails, the value of this
+// global indicates where the failure occurred.
+GC_LOAD_STATUS g_gc_load_status = GC_LOAD_STATUS_BEFORE_START;
+
+// The version of the GC that we have loaded.
+VersionInfo g_gc_version_info;
+
+class MyGCToEEInterface : public IGCToCLR {
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/clrgc.enabled.cpp]: We can use the same pattern as regular CoreCLR here. The lib split should not be a problem.

The current pattern used by regular CoreCLR has slightly better speed at the cost of duplicating native code. The current pattern used in this PR has slightly better code size, it does not duplicate native code but introduces extra method layer on GCtoEE calls. The difference is probably in the noise range.

If you like what you have in this PR better, we should unify on that regular CoreCLR. It can look like this: https://github.com/jkotas/runtime/commit/552e41bb8a055566394486e93da0cd8c0c3af8b0 . Saves ~200 lines of the boilerplate code.

File: src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt [hunk 16]
```diff
@@ -32,6 +32,9 @@ add_dependencies(Runtime.WorkstationGC aot_eventing_headers)
 add_library(Runtime.ServerGC STATIC ${COMMON_RUNTIME_SOURCES} ${FULL_RUNTIME_SOURCES} ${RUNTIME_SOURCES_ARCH_ASM} ${SERVER_GC_SOURCES} ${RUNTIME_ARCH_ASM_OBJECTS})
 add_dependencies(Runtime.ServerGC aot_eventing_headers)
 
+add_library(standalonegc-disabled STATIC ${CLRGC_DISABLED_SOURCES})
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt:35]: Nit: CLRGC can be renamed to STANDALONEGC here and the remaining places for consistency.

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

**Comment by lewing** [comment]
This is breaking the official build because of the new files

**Comment by lewing** [comment]
I opened https://github.com/dotnet/runtime/pull/93181 based on the error message
